### PR TITLE
Use RetryingHiveMetaStoreClient instead of HiveMetaStoreClient

### PR DESCRIPTION
**Pull request by suvodeep-pyne**
Azkaban currently uses `HiveMetaStoreClient` while communicating with the Hive Metastore ref class: `HadoopSecurityManager_H_2_0`
This client does not automatically retry/reconnect in case of transient failures, such as when a metastore is restarted. The RetryingHiveMetaStoreClient does automatically retry/reconnect in case of errors.

This change switches it to use the RetryingHiveMetaStoreClient (which is what the Hive CliDriver uses internally).

NOTE: This introduces a runtime dependency on `hive-exec` jar. Currently, this is set as a compile only dependency in Azkaban but it actually requires this jar in the runtime and expects to get it from the `HIVE_HOME` environment variable.

Ref: https://github.com/linkedin/gobblin/issues/1506

### CODE REVIEW

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    msc.dropDatabase(dbName, true, true);
+    msc.createDatabase(db);
+    db = msc.getDatabase(dbName);
+    Assert.assertEquals(db.getName(), dbName);
+    Assert.assertEquals(db.getDescription(), description);
+    Assert.assertEquals(db.getLocationUri(), location);
+
+    // Clean up if the test is successful
+    cleanup();
```
Comment (chengren311) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: maybe put it in finally block?

Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: I would prefer the cleanup to not take place if something goes wrong. That is the reason I didn't put it a `finally` block. Else, I would have simply used `@AfterTest` to ensure that the cleanup always takes place.

**Comment by suvodeep-pyne** [comment]
This is a smart catch by Travis. Apparently [this](https://mvnrepository.com/artifact/org.pentaho/pentaho-aggdesigner-algorithm/5.1.5-jhyde) dependency comes transitively and is not present in maven central but present in jcenter. This was cached on my machine but it fails on github.

File: azkaban-hadoop-security-plugin/build.gradle [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+
 dependencies {
     compile project(":azkaban-common")
 
     compileOnly deps.hadoopCommon
     compileOnly deps.hadoopMRClientCommon
     compileOnly deps.hadoopMRClientCore
     compileOnly deps.hiveMetastore
+    compileOnly(deps.hiveExec) {
+        exclude group: 'org.pentaho', module: 'pentaho-aggdesigner-algorithm'
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Why is it necessary to exclude these modules?

File: azkaban-hadoop-security-plugin/build.gradle [hunk 3]
```diff
@@ -1,12 +1,20 @@
 apply plugin: 'distribution'
 
+configurations {
+    testCompile.extendsFrom compileOnly
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Why is this change necessary?
How is it different from modifying testCompileOnly?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Nice. Thanks for remembering to add the notice. I configured Idea to add it automatically. Did you do the same?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link BasePooledObjectFactory} for {@link IMetaStoreClient}.
+ */
+public class HiveMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+  public static final String HIVE_METASTORE_TOKEN_SIGNATURE = "hive.metastore.token.signature";
+  private static final Logger LOG = LoggerFactory.getLogger(HiveMetaStoreClientFactory.class);
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: The convention is to use "logger". 
LOG should be marked as violating the google style guide.

Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Did you set up the checkstyle check? We should standardize the setup.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        return null;
+      }
+
+      try {
+        final HiveStorageHandler storageHandler =
+            HiveUtils.getStorageHandler(HiveMetaStoreClientFactory.this.hiveConf,
+                tbl.getParameters().get(META_TABLE_STORAGE));
+        return storageHandler == null ? null : storageHandler.getMetaHook();
+      } catch (final HiveException e) {
+        LOG.error(e.toString());
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Should log the stack too?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        throw new MetaException("Failed to get storage handler: " + e);
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: shouldn't it be , e to include the inner exception?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+    return RetryingMetaStoreClient
+        .getProxy(this.hiveConf, hookLoader, HiveMetaStoreClient.class.getName());
+  }
+
+  @Override
+  public IMetaStoreClient create() {
+    try {
+      return createMetaStoreClient();
+    } catch (final MetaException e) {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: If the exception is caught here, why not catch the HiveException here?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.hadoop.hive.ql.metadata.HiveStorageHandler;
+import org.apache.hadoop.hive.ql.metadata.HiveUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link BasePooledObjectFactory} for {@link IMetaStoreClient}.
+ */
+public class HiveMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Is this code new or is it moved from somewhere else?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 10]
```diff
@@ -402,11 +403,10 @@ private void getToken(final String userToProxy) throws InterruptedException,
     }
   }
 
-  private void cancelHiveToken(final Token<? extends TokenIdentifier> t,
-      final String userToProxy) throws HadoopSecurityManagerException {
+  private void cancelHiveToken(final Token<? extends TokenIdentifier> t)
+      throws HadoopSecurityManagerException {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: Is this an independent code clean up change?

File: build.gradle [hunk 11]
```diff
@@ -46,6 +46,7 @@ ext.deps = [
         hadoopMRClientCommon: "org.apache.hadoop:hadoop-mapreduce-client-common:" + versions.hadoop,
         hadoopMRClientCore  : "org.apache.hadoop:hadoop-mapreduce-client-core:" + versions.hadoop,
         hiveMetastore       : "org.apache.hive:hive-metastore:" + versions.hive,
+        hiveExec            : "org.apache.hive:hive-exec:" + versions.hive,
```
Comment (HappyRay) [inline build.gradle]: Shouldn't it come before hiveMetastore?

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: Why is this approach better?

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    final String dbName = "test_db";
+    final String description = "test database";
+    final String location = "file:/tmp/" + dbName;
+    Database db = new Database(dbName, description, location, null);
+
+    msc.dropDatabase(dbName, true, true);
+    msc.createDatabase(db);
+    db = msc.getDatabase(dbName);
+    Assert.assertEquals(db.getName(), dbName);
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: Consider using assertJ. 
It would be good that new code standardizes on assertJ.

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: I saw you had a cleanup call in the beginning of the test already. Is it the idea to be able to inspect the state after the test fails and still be able to run the next test run with a clean state?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+    final HiveConf hiveConf = new HiveConf();
+    hiveConf.setVar(HiveConf.ConfVars.METASTOREURIS, hcatURI);
+    hiveConf.set(HIVE_METASTORE_TOKEN_SIGNATURE, hcatURI);
+
+    return hiveConf;
+  }
+
+  private IMetaStoreClient createMetaStoreClient() throws MetaException {
+    final HiveMetaHookLoader hookLoader = tbl -> {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I am not familiar with this hive code. Is there a common utility function or an overload that will do this?  It doesn't look like it is specific to our use case.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+/**
+ * An implementation of {@link BasePooledObjectFactory} for {@link IMetaStoreClient}.
+ */
+public class HiveMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+  public static final String HIVE_METASTORE_TOKEN_SIGNATURE = "hive.metastore.token.signature";
+  private static final Logger LOG = LoggerFactory.getLogger(HiveMetaStoreClientFactory.class);
+  private final HiveConf hiveConf;
+
+  public HiveMetaStoreClientFactory() {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Why having three overloaded constructors when we only use one?

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java (hunk 1 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: @HappyRay yes. That was my motivation.

File: azkaban-hadoop-security-plugin/build.gradle (hunk 2 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/build.gradle]: Main cause: They are not present in maven central or jcenter.
Why we don't care? These modules are not required to compile the code or run the test cases.

File: azkaban-hadoop-security-plugin/build.gradle (hunk 3 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/build.gradle]: I don't want to duplicate each dependency. IMO, this is better than maintaining 2 duplicate dependency lists. That exposes us to cases like what if somebody adds it to one and not the other for example.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 4 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: yes :) . It adds the copyright notice whenever I create a new class.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 5 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Where is the convention mentioned? I use `log`. I find `log` to be shorter and yet equally expressive as `logger` but that's just my opinion. In that case the style guide should fail the build.

> Did you set up the checkstyle check?

Yes. I followed the development instructions around a week back. Did things change again?

Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: This class was picked up from the Gobblin code base. Let me refactor the code a bit.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 10 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: yes.

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java (hunk 12 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: Will take a look and update the PR.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 13 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I am not sure about this. This was the suggestion from our internal teams to use this implementation. For now, we can go ahead with this and we can refactor if we find something that is easier.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 14 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: borrowed this code from Gobblin. Thought it may be useful since we have the same use case. but I agree this is more confusion than help. I'll get rid of them

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 8 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: The `HiveException` is inside a Lambda. I don't think you can propagate that directly. Correct me if I'm wrong here.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 9 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: See the description. The code has been copied from Gobblin.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 5 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Regarding the `log` convention, I would just suggest using simply `log`
This is how lombok does it. link: https://projectlombok.org/features/log

**Comment by HappyRay** [comment]
How can these dependencies be cached on your machine if they are not available in the repository?

File: azkaban-hadoop-security-plugin/build.gradle (hunk 2 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: This is odd. Did you check with Hive team?

Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Consider adding a comment in the code if this change needs to stay.

File: azkaban-hadoop-security-plugin/build.gradle (hunk 3 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Don't you have the same problem if someone adds to compile a dependency this test module needs?

IMHO, it is better to stay with the default/convention to avoid surprises as much as possible.

If someone forgets to add a dependency, tests will fail. Right? Why isn't it good enough?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 10 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: I'd suggest you keep them separate in the future. It makes review harder. If you want a quicker review, keep the change as small and focused as possible.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 9 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I saw the link to Gobblin. But I didn't click through.

Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: It's not clear that Gobblin has to make these build changes. Should you have @erwa review this change?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 8 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Do you have a reference to this behavior ( a test program, article etc?)

If this is the issue, how does "  throw new MetaException("Failed to get storage handler: " + e);" work then?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+public class HiveMetaStoreClientFactory extends BasePooledObjectFactory<IMetaStoreClient> {
+
+  private static final Logger log = LoggerFactory.getLogger(HiveMetaStoreClientFactory.class);
+  private final HiveConf hiveConf;
+
+  public HiveMetaStoreClientFactory(final HiveConf hiveConf) {
+    this.hiveConf = hiveConf;
+  }
+
+  private IMetaStoreClient createMetaStoreClient() throws MetaException {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I don't follow why this code is needed and what it is doing. 
Consider adding comments.

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.thrift.TException;
+import org.junit.Test;
+
+public class HiveMetaStoreClientFactoryTest {
+
+  public static final File METASTORE_DB_DIR = new File("metastore_db");
+  public static final File DERBY_LOG_FILE = new File("derby.log");
+
+  @Test
+  public void testCreate() throws TException, IOException {
```
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: How does the test verify the retry behavior?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 5 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Either log or logger sounds good to me. The main problem is that it should be lower case. You can read the Google style doc. 
No, we haven't setup the proper style check instruction yet. Are you interested in helping?

Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: How does Gobblin enforce code styles?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 6 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Is the question clear?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 7 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: ?

File: azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java (hunk 12 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/test/java/azkaban/hive/HiveMetaStoreClientFactoryTest.java]: Thanks for making the update. It will make the job of the reviewers easier if you leave a comment suggesting that the change is made. Otherwise, the reviewers may have a hard time remembering which ones have been addressed.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 13 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: Please follow up. Simpler code is better code. 
And we need to understand what the code does.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 6 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I saw the change now. This is an example of why I found it hard to track which comment has been addressed. I wish Github has a feature like the Review Board has that can make it easier. Do you know such a feature in Github?

File: azkaban-hadoop-security-plugin/build.gradle (hunk 3 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/build.gradle]: No, I don't think so. `testCompile` extends `compile` by default. But it doesn't seem to pull in from `compileOnly` which is why I added this line.

> IMHO, it is better to stay with the default/convention to avoid surprises as much as possible.

I find it intuitive to have the compileOnly classes available in my test classpath so that I can test my code! In fact, it is counter intuitive to me if my jar gets built successfully yet my tests don't compile.

Gradle configurations are built from file collections so that they can be extended, filtered and manipulated. So I don't think this should come as a surprise to a Gradle user. In fact, I think it might help to take a look at the task and configuration graph.
link: https://docs.gradle.org/3.3/userguide/java_plugin.html

> If someone forgets to add a dependency, tests will fail. Right? Why isn't it good enough?

This is good. But why do have to add it in 2 places is the question.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 8 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: The `MetaException` is part of the interface which the Lambda accepts. Pasting the interface below. 

```

/**
 * HiveMetaHookLoader is responsible for loading a {@link HiveMetaHook}
 * for a given table.
 */
public interface HiveMetaHookLoader {
  /**
   * Loads a hook for the specified table.
   *
   * @param tbl table of interest
   *
   * @return hook, or null if none registered
   */
  public HiveMetaHook getHook(Table tbl) throws MetaException;
}
```

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 9 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: I am not too concerned about this since it is simply instantiating a HMS client. But I'm totally open to getting some feedback on how we can do this better.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 10 above)
Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: I agree this is off context. I tried the other way and I see myself making less changes overall and it's counter productive for me. So, I would rather choose to make the code better and wait a little longer for the review. I tend to switch to other things and often forget / feel lazy to create a refactor patch before/after my code changes. 

I don't want to make a whole bunch of changes as well but definitely the ones that come to my eye. This one, for example, has been lurking around for a long time and I haven't touched it because I wanted to make a refactor PR that never happened.

**Comment by suvodeep-pyne** [comment]
@HappyRay 
> How can these dependencies be cached on your machine if they are not available in the repository?
I remember making some extra effort in fetching dependencies from elsewhere. This was when I was working on the Hive codebase.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 8 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: This just indicates that this method will throw this checked exception. Will it work if we change it to throws HiveExecption?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java (hunk 9 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/hive/HiveMetaStoreClientFactory.java]: @erwa could you comment?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 10 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: "This one, for example, has been lurking around for a long time and I haven't touched it because I wanted to make a refactor PR that never happened." How can I help so that you can make it happen more often in the future? :) Any suggestion?

**Comment by HappyRay** [comment]
Discussed off line, due to the risk of not having this fix, we agreed to check in this fix first. @suvodeep-pyne agreed to address the remaining review comments later.

**Comment by HappyRay** [comment]
The test took almost 9 seconds to run on my mac. This is slow. Any idea how to speed it up?

**Comment by HappyRay** [comment]
How did you test this feature work?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java (hunk 10 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java]: I would take into consideration of the reviewers and more importantly the productivities of the future readers of this code and change history.

File: azkaban-hadoop-security-plugin/build.gradle (hunk 3 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Modifying the sourceSets is not an issue and is expected. 
I thought extendsFrom would **replace** the existing superConfigs. After some experiments, I realized that it **adds** to the existing configs. 
This change is fine. 
Also since the build only packages the jar and not the run time dependencies, does it need to use compileOnly in the first place? What's your long term plan?

Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/build.gradle]: > Modifying the sourceSets is not an issue and is expected.

AFAIK, I didn't modify any source set. I don't understand the context here.

> I thought extendsFrom would replace the existing superConfigs. After some experiments, I realized that it adds to the existing configs.

Yes. It simply _extends_ the configuration/collection with another.

> Also since the build only packages the jar and not the run time dependencies, does it need to use compileOnly in the first place? What's your long term plan?

Correction: build packages the jar WITH the runtime dependencies. `compileOnly` means that this collection is only used for compiling the artifact and they are not added to runtime. It does not need to be `compileOnly` since HDFS storage introduced Hadoop as a compile level dependency at this time. But prior to that, Hadoop and Hive dependencies were not shipped with Azkaban.

The long term plan IMO should be to get rid of Hadoop, Hive etc dependencies from core Azkaban. HDFS Storage and Hadoop Security Manager code should run in a plugin module like environment.

Comment (suvodeep-pyne) [inline azkaban-hadoop-security-plugin/build.gradle]: @HappyRay 
Let me know if we should do probably a 30-min or 1-hour discussion on Gradle some time since a GitHub PR doesn't seem to be well suited for explaining Gradle intricacies.

**Comment by suvodeep-pyne** [comment]
> The test took almost 9 seconds to run on my mac. This is slow. Any idea how to speed it up?

It is taking some seconds to run since it sets up a DB and brings up the Hive Metastore. One way of speeding it up would be to remove the cleanup before running the test. This would benefit second runs in which the HMS start up wouldn't have to create the local files every time.

**Comment by suvodeep-pyne** [comment]
> How did you test this feature work?

I didn't test whether `RetryingHiveMetaStoreClient` works or not. This is a Hive class. I just tested whether existing Hive jobs are running fine. This would ensure that they are fetching the delegation tokens correctly.

File: azkaban-hadoop-security-plugin/build.gradle (hunk 3 above)
Comment (HappyRay) [inline azkaban-hadoop-security-plugin/build.gradle]: Sounds good.
Sorry, I meant configurations, not sourceSet.

**Comment by HappyRay** [comment]
That test is good for catching regressions. But it doesn't test the new functionality is working, does it?
